src/TorchSharp/NN/Module.cs

```diff
@@ -1148,21 +1149,120 @@ public interface IModule<T1, T2, T3, TResult>
+            }
+
+            public class HookableModule<TPreHook,TPostHook> : Module
+            {
+                protected HookableModule(string name) : base(name) { }
+
+                protected HookableModule(IntPtr handle, IntPtr boxedHandle) : base(handle, boxedHandle) { }
+                internal HookableModule(HType handle, IntPtr? boxedHandle) : base(handle, boxedHandle) { }
+
+                public HookRemover register_forward_hook(TPostHook hook)
+                {
+                    var key = Guid.NewGuid().ToString();
+                    post_hooks.Add(key, hook);
+                    return new HookRemover(this, key);
+                }
+
+                public HookRemover register_forward_pre_hook(TPreHook hook)
+                {
+                    var key = Guid.NewGuid().ToString();
+                    pre_hooks.Add(key, hook);
+                    return new HookRemover(this, key);
+                }
+
+                private void remove(string key)
+                {
+                    if (pre_hooks.ContainsKey(key)) pre_hooks.Remove(key);
+                    if (post_hooks.ContainsKey(key)) post_hooks.Remove(key);
+                }
+
+                protected Dictionary<string, TPreHook> pre_hooks = new Dictionary<string, TPreHook>();
+                protected Dictionary<string, TPostHook> post_hooks = new Dictionary<string, TPostHook>();
+
+                public class HookRemover
+                {
+                    public HookRemover(HookableModule<TPreHook, TPostHook> module, string key)
+                    {
+                        this.module = module;
+                        this.key = key;
+                    }
+
+                    public void remove()
+                    {
+                        module.remove(key);
+                    }
+
+                    private HookableModule<TPreHook, TPostHook> module;
+                    private string key;
+                }
+            }
 
             /// <summary>
             /// Base class for concrete modules with a forward() that takes a single argument.
             /// </summary>
             /// <typeparam name="T">The argument type of the module's forward() function.</typeparam>
             /// <typeparam name="TResult">The return type of the module's forward() function.</typeparam>
-            public abstract class Module<T, TResult> : Module, IModule<T, TResult>
+            public abstract class Module<T, TResult> : HookableModule<Func<Module<T,TResult>, T, T>, Func<Module<T, TResult>, T, TResult, TResult>>, IModule<T, TResult>
             {
                 protected Module(string name) : base(name) { }
                 protected Module(IntPtr handle, IntPtr boxedHandle) : base(handle, boxedHandle) { }
                 internal Module(HType handle, IntPtr? boxedHandle) : base(handle, boxedHandle) { }
-                public abstract TResult forward(T input1);
+                protected abstract TResult forward(T input1);
```

Comment:
I think this is okay, but the user can call `forward()` in PyTorch, you cannot with this change.

Reply:
Yes, I was thinking about that. I see some pros and cons with having 'forward' be public:

1. It's confusing to have two ways to call the module -- only one of which is hookable.
2. Existing models wouldn't be forced to change, so hooking would only be possible for new models that are explicitly set up for it.
3. On the other hand, keeping forward as a public will avoid a breaking change.

So, I'm not sure which is better.

Reply:
Some pytorch tutorial calls forward()

Reply:
If you're worried about confusion, why not simply document the differences between the two ways you can call the module and call it a day? Breaking changes without a strict technical requirement are the worst.

Reply:
@kaiidams, @JustinWick, @dayo05 -- I'm persuaded. The Python behavior is actually that the hooks don't get called when you call 'm.forward(x)` instead of `m(x)`. So, this will avoid a breaking change and works more like Pytorch.

I'll make `forward()` public.